SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/LayeredLBStrategy.cpp

```diff
@@ -0,0 +1,278 @@
+			AddStrategyForLayer(SpatialConstants::DefaultLayer, DefaultLBStrategy);
+			return;
+		}
+
+		// This will be uncommented after the next PR.
+		// For each Layer, add a LB Strategy for that layer.
+// 			TMap<FName, FLBLayerInfo> WorkerLBLayers = WorldSettings->WorkerLBLayers;
+// 
+// 			for (const TPair<FName, FLayerInfo>& Layer : Settings->WorkerLayers)
+// 			{
+// 				FName LayerName = Layer.Key;
+// 
+// 				// Look through the WorldSettings to find the LBStrategy type for this layer.
+// 				if (!WorkerLBLayers.Contains(LayerName))
+// 				{
+// 					UE_LOG(LogLayeredLBStrategy, Error, TEXT("Layer %s does not have a defined LBStrategy in the WorldSettings. It will not be simulated."), *(LayerName.ToString()));
+// 					continue;
+// 				}
+// 
+// 				UAbstractLBStrategy* LBStrategy = NewObject<UAbstractLBStrategy>(this, WorkerLBLayers[LayerName].LoadBalanceStrategy);
+// 				AddStrategyForLayer(LayerName, LBStrategy);
+// 
+// 				for (const TSoftClassPtr<AActor>& ClassPtr : Layer.Value.ActorClasses)
+// 				{
+// 					ClassPathToLayer.Add(ClassPtr, LayerName);
+// 				}
+// 			}
+
+		// Finally, add the default layer.
+		if (WorldSettings->LoadBalanceStrategy == nullptr)
+		{
+			UE_LOG(LogLayeredLBStrategy, Error, TEXT("If EnableUnrealLoadBalancer is set, there must be a LoadBalancing strategy set. Using a 1x1 grid."));
+			UAbstractLBStrategy* DefaultLBStrategy = NewObject<UGridBasedLBStrategy>(this);
+			AddStrategyForLayer(SpatialConstants::DefaultLayer, DefaultLBStrategy);
+		}
+		else
+		{
+			UAbstractLBStrategy* DefaultLBStrategy = NewObject<UAbstractLBStrategy>(this, WorldSettings->LoadBalanceStrategy);
+			AddStrategyForLayer(SpatialConstants::DefaultLayer, DefaultLBStrategy);
+		}
+	}
+}
+
+void ULayeredLBStrategy::SetLocalVirtualWorkerId(VirtualWorkerId InLocalVirtualWorkerId)
+{
+	LocalVirtualWorkerId = InLocalVirtualWorkerId;
+	for (const auto& Elem : LayerNameToLBStrategy)
+	{
+		Elem.Value->SetLocalVirtualWorkerId(InLocalVirtualWorkerId);
+	}
+}
+
+bool ULayeredLBStrategy::ShouldHaveAuthority(const AActor& Actor) const
+{
+	if (!IsReady())
+	{
+		UE_LOG(LogLayeredLBStrategy, Warning, TEXT("LayeredLBStrategy not ready to relinquish authority for Actor %s."), *AActor::GetDebugName(&Actor));
+		return false;
+	}
+
+	const FName& LayerName = GetLayerNameForActor(Actor);
+	if (!LayerNameToLBStrategy.Contains(LayerName))
+	{
+		UE_LOG(LogLayeredLBStrategy, Warning, TEXT("LayeredLBStrategy doesn't have a LBStrategy for Actor %s which is in Layer %s."), *AActor::GetDebugName(&Actor), *LayerName.ToString());
+		return false;
+	}
+
+	// If this worker is not responsible for the Actor's layer, just return false.
+	if (VirtualWorkerIdToLayerName.Contains(LocalVirtualWorkerId) && VirtualWorkerIdToLayerName[LocalVirtualWorkerId] != LayerName)
+	{
+		return false;
+	}
+
+	return LayerNameToLBStrategy[LayerName]->ShouldHaveAuthority(Actor);
+}
+
+VirtualWorkerId ULayeredLBStrategy::WhoShouldHaveAuthority(const AActor& Actor) const
```

Comment:
Lot of duplication here, can `ShouldHaveAuthority` call `WhoShouldHaveAuthority` and check the result against its local id?

Reply:
It is a lot of duplication, but ShouldHaveAuthority gets to skip calling into the underlying LBStrategy layerCount-1/layerCount % of the time because it early-outs if the Layer doesn't match, so there is a possibly significant performance gain.